civis/tests/test_parallel.py

```diff
@@ -482,6 +482,22 @@ def test_result_eventual_failure(mock_civis):
     assert callback.call_count == 0
 
 
+@mock.patch.object(civis.parallel, 'civis')
+def test_result_running_and_cancel_requested(mock_civis):
+    callback = mock.MagicMock()
+    mock_civis.io.civis_to_file.side_effect = make_to_file_mock('spam')
+    response = Response({'is_cancel_requested': True,
+                         'state': 'running'})
+    client = mock.MagicMock()
+    client.scripts.post_cancel.return_value = response
+    fut = ContainerFuture(1, 2, client=client)
+    fut.set_result(response)
+    civis.parallel._CivisBackendResult(fut, callback)
```

Comment:
Is this to emulate what the joblib backend does? Would it be too difficult to mock out parts of the joblib backend and test with that?

Reply:
More or less yeah. Since the bug was in _CivisBackendResult, I was trying isolate it as closely as possible by forcing the future callback to be called with state=running and is_cancel_requested=True, then checking to make sure the joblib callback hadn't been called. I could also have mocked the joblib but this was the first test that came to mind. Do you think it'd be worth it to test _CivisBackend?

Reply:
I think this is probably fine as a test of the functionality, though a test using _CivisBackend would be easier to understand, I think.

If you want to keep this approach, I think it's a bit confusing to look at this test code right now because it's not obvious how the `cancel` call would affect the callback, so I think it'd benefit from a comment. How about something like this?

"When a future created by the Civis joblib backend completes normally, a callback function will be executed. If the future is cancelled, we don't want that callback to be run. Instantiating the _CivisBackendResult object emulates what the backend does internally."

Reply:
👍 Good idea, added some explanation